### PR TITLE
SONARJAVA-2095: Avoid creating temporary lists when inspecting stack

### PR DESCRIPTION
**Pull request by saberduck**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/se/ProgramState.java [hunk 1]
```diff
@@ -417,4 +422,9 @@ public SymbolicValue exitValue() {
   public boolean exitingOnRuntimeException() {
     return exitSymbolicValue instanceof SymbolicValue.ExceptionalSymbolicValue && ((SymbolicValue.ExceptionalSymbolicValue) exitSymbolicValue).exceptionType() == null;
   }
+
+  public boolean stackIsEmpty() {
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/ProgramState.java]: I'm not convinced by the interest of this one method. You are basically creating it only to use it in one place. See other comment.

File: java-frontend/src/main/java/org/sonar/java/se/checks/NullDereferenceCheck.java [hunk 2]
```diff
@@ -43,26 +43,25 @@
 
   @Override
   public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
-    SymbolicValue currentVal = context.getState().peekValue();
-    if (currentVal == null) {
+    if (context.getState().stackIsEmpty()) {
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/checks/NullDereferenceCheck.java]: Why not simply create a local method `stackIsEmpty(ProgramState state)` doing the null-check?
As it's the only place you need it, creating a method in ProgramState seems over-engineering to me.

File: java-frontend/src/main/java/org/sonar/java/se/checks/NullDereferenceCheck.java [hunk 3]
```diff
@@ -43,26 +43,25 @@
...
     if (syntaxNode.is(Tree.Kind.METHOD_INVOCATION)) {
       MethodInvocationTree methodInvocation = (MethodInvocationTree) syntaxNode;
-      toCheck = methodInvocation.methodSelect();
-      int numberArguments = methodInvocation.arguments().size();
-      List<SymbolicValue> values = context.getState().peekValues(numberArguments + 1);
-      currentVal = values.get(numberArguments);
+      Tree methodSelect = methodInvocation.methodSelect();
+      if (methodSelect.is(Tree.Kind.MEMBER_SELECT)) {
+        SymbolicValue dereferencedSV = context.getState().peekValue(methodInvocation.arguments().size());
+        return checkConstraint(context, methodSelect, dereferencedSV);
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/checks/NullDereferenceCheck.java]: nice improvement :+1:

File: java-frontend/src/main/java/org/sonar/java/collections/PStack.java [hunk 4]
```diff
@@ -41,6 +43,14 @@
   E peek();
 
   /**
+   *
+   * @param i - index of element to be returned, 0 means top of the stack
+   * @return i-th element from top of the stack or null, if stack doesn't have so many elements
+   */
+  @CheckForNull
+  E peek(int i);
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/collections/PStack.java]: I would prefer to have the method throw an `IllegalStateException` if we try to access something out of the stack. When using this method, it means that you perfectly know what you are doing, trying to access a given element from the stack. If you can't I would prefer to have it failing fast.

File: java-frontend/src/main/java/org/sonar/java/se/ProgramState.java (hunk 1 above)
Comment (saberduck) [inline java-frontend/src/main/java/org/sonar/java/se/ProgramState.java]: I removed the method for now, but I think I will reintroduce it once we make also ProgramState.peek to throw an exception

File: java-frontend/src/main/java/org/sonar/java/se/checks/NullDereferenceCheck.java (hunk 2 above)
Comment (saberduck) [inline java-frontend/src/main/java/org/sonar/java/se/checks/NullDereferenceCheck.java]: see first comment

**Pull request merged**